### ISSUE TITLE
ScmDataFrame use cases

### ISSUE THREAD

**Issue by znicholls** [openclimatedata/openscm#169]
@swillner as promised, here are some thoughts to start the discussion.

The basic pipeline I envisage is this.

1. The user has some data they want to run in OpenSCM
1. They convert this data into an ScmDataFrame instance
    - at this point there is no `climate_model` column
1. They pass the ScmDataFrame instance to some sort of `run` function, along with a set of climate model and configurations to run (at the moment I was thinking this would be a `dict` or `climate_model`-`ParameterSet` key-value pairs but we could do whatever)
1. The `run` function loops over the climate models to run, for each climate_model
    1. loop over unique scenarios in the `ScmDataFrame`, for each scenario
        1. convert the scenario to a parameterset
            - the set of timeseries should be unique for each scenario i.e. a scenario wouldn't define e.g. 'Emissions|CO2' twice
            - this is where we require `to_core` or `to_parameterset`
            1. initialise the climate_model's inputs using the scenario parameterset
            1. for each configuration
                1. initialise the climate_model's run_parameters
                1. run the model
                1. store the outputs
                    - we could call `to_scmdataframe` here (the output ScmDataFrame would have a `climate_model` column to identify which climate model was run as a shortcut from the user so they don't have to work it out by looking at all the other parameters)
1. collect all the stored outputs (each being a unique scenario-climate_model-parameter combination) into a single ScmDataFrame
    - we could instead call `to_scmdataframe` here (the output ScmDataFrame would have a `climate_model` column to identify which climate model was run as a shortcut from the user so they don't have to work it out by looking at all the other parameters)
1. do whatever analysis on the outputs, using ScmDataFrame's analysis capability (the `climate_model` column comes in very handy here)

**Comment by swillner** [comment]
sounds good. from the outside this should look like (with ScmDataFrame `df`):
```
from openscm import Ensemble

ensemble = Ensemble(df, ["PH99", "MAGICC6", "Hector"])
ensemble.run(start_time, stop_time)
```
Can have a go at the implementation beginning of next week. Would need to overhaul `convert_openscm_to_scmdataframe` for that.

A single model run should look prett much the same:
```
from openscm import OpenSCM

model = OpenSCM("PH99", input_parameters=df.to_parameterset())
# could alter/add parameters here
model.run(start_time, stop_time)
model.results_to_scmdataframe(df)
```
(which would be used by the `Ensemble` class internally).

@znicholls @rgieseke What do you think?

**Comment by swillner** [comment]
@znicholls regarding https://github.com/znicholls/openscm/blob/run-function/notebooks/pyam-integration.ipynb:
- We should not and do not need to use pint units directly (as per our documentation there)
- Still not sure if and how we should handle start_time and stop_time in the ParameterSet
- I think running a model should look like above

**Comment by znicholls** [comment]
>     * We should not and do not need to use pint units directly (as per our documentation there)

yep lazy hack by me

>     * Still not sure if and how we should handle start_time and stop_time in the ParameterSet

I'd put them in `ParameterSet`, models can set their defaults easily that way

> What do you think?

I like it, with minor changes

```python
from openscm import Ensemble

# I'd allow the user to also pass model specific parameters into `Ensemble` 
# so would have a dict rather than a list (but happy with other solutions too)
ensemble = Ensemble(df, {"PH99": ph99_specfic_pars, "MAGICC6": magicc6_specific_paras})
ensemble.run()  # start and stop time go in parameter sets
```

```python
from openscm import OpenSCM

model = OpenSCM("PH99", input_parameters=df.to_parameterset())
# could alter/add parameters here
model.run()  # start and stop time go in parameter sets
out_df = model.results_to_scmdataframe()  # don't alter dataframes inplace by default, it's a nightmare
```

**Comment by swillner** [comment]
> don't alter dataframes inplace by default, it's a nightmare

so lose input parameters then? why is there no easy way to just add some rows?

**Comment by swillner** [comment]
for the ensemble parameters:
Why model-specific? In an ensmble all should be the same and adapters ignore unknown parameters anyway... so also using a shared parameterset?

**Comment by znicholls** [comment]
> so lose input parameters then? why is there no easy way to just add some rows?

No, I would think model results would also return the inputs. It's a weird construction to pass a dataframe to a function, return nothing but have the dataframe be altered. I'd really prefer to avoid that.

> adapters ignore unknown parameters anyway

I always forget this. On the same page now.

**Comment by swillner** [comment]
> No, I would think model results would also return the inputs. It's a weird construction to pass a dataframe to a function, return nothing but have the dataframe be altered. I'd really prefer to avoid that.

could add a `add_results` function to ScmDataFrame?

**Comment by znicholls** [comment]
Nice solution!
